### PR TITLE
frontend: migrate onto esbuild for dev builds

### PR DESCRIPTION
**Pull request by dschaller**
### Description

Migrates off of babel for the development workflow onto esbuild.

This also modifies the lint job to use esbuild to compile components.

### GitHub Issue

Fixes #

### TODOs

- [ ] Figure out a better way to distribute dev build script from tooling

### CODE REVIEW

File: .github/workflows/frontend.yml [hunk 1]
```diff
@@ -81,26 +85,68 @@ jobs:
...
+        id: yarn-cache
+        run: echo "::set-output name=dir::$(yarn cache dir)"
+      - uses: actions/cache@v2
+        with:
+          path: ${{ steps.yarn-cache.outputs.dir }}
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+      - name: Install dependencies
+        if: steps.cache.outputs.cache-hit != 'true'
```
Comment (mcutalo88) [inline .github/workflows/frontend.yml:140]: I'm not sure how this fully works but should this be `steps.yarn-cache`? Are the keys in this statement here the `id` of each step?

Comment (dschaller) [reply .github/workflows/frontend.yml:140]: oh yeah good catch

File: frontend/packages/tools/devBuild.mjs [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import esbuild from "esbuild";
+import fs from "fs";
+import path from "path";
+
+const args = process.argv.slice(2);
```
Comment (mcutalo88) [inline frontend/packages/tools/devBuild.mjs:5]: whats the `slice(2)` for?

File: frontend/workflows/envoy/build.mjs [hunk 3]
```diff
@@ -0,0 +1,32 @@
+import esbuild from "esbuild";
```
Comment (mcutalo88) [inline frontend/workflows/envoy/build.mjs:1]: Should we add this file to the workflow scaffolder in this pr?

File: .github/workflows/frontend.yml [hunk 4]
```diff
@@ -81,26 +85,68 @@ jobs:
...
+          restore-keys: |
+            ${{ runner.os }}-go-mod-
+      - name: go build cache
+        uses: actions/cache@v2
+        with:
+          path: |
+            ~/.cache/go-build
+          key: ${{ runner.os }}-go-build-${{ hashFiles('**/go.sum') }}
+          restore-keys: |
+            ${{ runner.os }}-go-build-
```
Comment (danielhochman) [inline .github/workflows/frontend.yml:103]: so we dont cache the smaller build cache from the mock server for the regular build
```suggestion
          key: ${{ runner.os }}-go-mock-build-${{ hashFiles('**/go.sum') }}
          restore-keys: |
            ${{ runner.os }}-go-build-
            ${{ runner.os }}-go-mock-build-
```

File: frontend/workflows/sourcecontrol/package.json [hunk 5]
```diff
@@ -12,8 +12,8 @@
     "build": "yarn clean && yarn compile",
     "clean": "rm -rf ./dist && rm -f tsconfig.tsbuildinfo",
     "compile": "tsc -b",
-    "compile:dev": "tsc --sourceMap",
-    "compile:watch": "BABEL_ENV=build babel src --out-dir dist --source-maps --extensions .js,.jsx,.ts,.tsx --ignore **/tests --watch",
+    "compile:dev": "esbuild --target=es2019 --outdir=dist --sourcemap src/*.tsx",
+    "compile:watch": "esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx",
```
Comment (danielhochman) [inline frontend/workflows/sourcecontrol/package.json:16]: why limit to 2019?

Comment (dschaller) [reply frontend/workflows/sourcecontrol/package.json:16]: This is to enable support for optional chaining

File: frontend/workflows/envoy/build.mjs (hunk 3 above)
Comment (dschaller) [reply frontend/workflows/envoy/build.mjs:1]: I'd like to figure out how to package this script in a way where it's easier to distribute.

File: frontend/packages/tools/devBuild.mjs (hunk 2 above)
Comment (dschaller) [reply frontend/packages/tools/devBuild.mjs:5]: the second command line argument (e.g. in this case it's the source directory)

**Comment by danielhochman** [comment]
The default is esnext which should cover all the latest JS features. I
believe it targets es2020 for now.

File: frontend/workflows/sourcecontrol/package.json (hunk 5 above)
Comment (dschaller) [reply frontend/workflows/sourcecontrol/package.json:16]: >The default is esnext which should cover all the latest JS features. I
believe it targets es2020 for now.

Huh I swear I tried this and it didn't work over the weekend but magically it does now 🤔

**Comment by danielhochman** [comment]
ctrl+c in the middle of a build left a ton of leftover processes for me. we might need a wrapper that tracks all child PIDs, catches signal, and kills if the tools don't handle them properly in the first place.

```
dhochman  427036  426968  0 Mar15 pts/0    00:00:00 /bin/sh -c esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx
dhochman  427041  427036  2 Mar15 pts/0    01:20:42 esbuild --target=es2019 --outdir=dist --sourcemap --watch src/index.tsx src/reboot-instance.tsx src/resize-asg.tsx src/terminate-instance.tsx
dhochman  427049  426956  0 Mar15 pts/0    00:00:00 /bin/sh -c yarn workspace @clutch-sh/tools buildDev $PWD -w
dhochman  427050  426971  0 Mar15 pts/0    00:00:00 /bin/sh -c node build.mjs $PWD -w
dhochman  427052  427049  0 Mar15 pts/0    00:00:31 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js workspace @clutch-sh/tools buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/data-layout -w
dhochman  427053  427050  0 Mar15 pts/0    00:00:45 /usr/bin/node build.mjs /home/dhochman/go/src/github.com/lyft/clutch/frontend/workflows/envoy -w
dhochman  427060  426989  0 Mar15 pts/0    00:00:00 /bin/sh -c esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx
dhochman  427061  427060  2 Mar15 pts/0    01:22:00 esbuild --target=es2019 --outdir=dist --sourcemap --watch src/form-fields.tsx src/index.tsx src/start-experiment.tsx
dhochman  427065  426961  0 Mar15 pts/0    00:00:00 /bin/sh -c yarn workspace @clutch-sh/tools buildDev $PWD -w
dhochman  427074  427065  0 Mar15 pts/0    00:00:31 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js workspace @clutch-sh/tools buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/wizard -w
dhochman  427084  426950  0 Mar15 pts/0    00:00:00 /bin/sh -c yarn workspace @clutch-sh/tools buildDev $PWD -w
dhochman  427085  427084  0 Mar15 pts/0    00:00:30 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js workspace @clutch-sh/tools buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/core -w
dhochman  427107  426979  0 Mar15 pts/0    00:00:00 /bin/sh -c node build.mjs $PWD -w
dhochman  427109  426981  0 Mar15 pts/0    00:00:00 /bin/sh -c esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx
dhochman  427110  427053  3 Mar15 pts/0    01:43:46 /home/dhochman/go/src/github.com/lyft/clutch/frontend/node_modules/esbuild/bin/esbuild --service=0.9.2 --ping
dhochman  427111  427109  2 Mar15 pts/0    01:21:44 esbuild --target=es2019 --outdir=dist --sourcemap --watch src/delete-pod.tsx src/index.tsx src/resize-hpa.tsx
dhochman  427116  427107  0 Mar15 pts/0    00:00:45 /usr/bin/node build.mjs /home/dhochman/go/src/github.com/lyft/clutch/frontend/workflows/experimentation -w
dhochman  427132  426983  0 Mar15 pts/0    00:00:00 /bin/sh -c esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx
dhochman  427133  427132  2 Mar15 pts/0    01:19:16 esbuild --target=es2019 --outdir=dist --sourcemap --watch src/index.tsx src/update-shard-count.tsx
dhochman  427151  427005  0 Mar15 pts/0    00:00:00 /bin/sh -c esbuild --target=es2019 --outdir=dist --sourcemap --watch src/*.tsx
dhochman  427152  427151  2 Mar15 pts/0    01:20:00 esbuild --target=es2019 --outdir=dist --sourcemap --watch src/create-repository.tsx src/index.tsx
dhochman  427161  427052  0 Mar15 pts/0    00:00:31 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/data-layout -w
dhochman  427167  427116  3 Mar15 pts/0    01:43:33 /home/dhochman/go/src/github.com/lyft/clutch/frontend/node_modules/esbuild/bin/esbuild --service=0.9.2 --ping
dhochman  427178  427085  0 Mar15 pts/0    00:00:30 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/core -w
dhochman  427185  427074  0 Mar15 pts/0    00:00:30 /usr/bin/node /home/dhochman/go/src/github.com/lyft/clutch/frontend/.yarn/releases/yarn-1.22.5.js buildDev /home/dhochman/go/src/github.com/lyft/clutch/frontend/packages/wizard -w
dhochman  427220  427200  3 Mar15 pts/0    01:44:56 /home/dhochman/go/src/github.com/lyft/clutch/frontend/node_modules/esbuild/bin/esbuild --service=0.9.2 --ping
dhochman  427246  427225  3 Mar15 pts/0    01:34:18 /home/dhochman/go/src/github.com/lyft/clutch/frontend/node_modules/esbuild/bin/esbuild --service=0.9.2 --ping
dhochman  427260  427239  3 Mar15 pts/0    01:43:21 /home/dhochman/go/src/github.com/lyft/clutch/frontend/node_modules/esbuild/bin/esbuild --service=0.9.2 --ping
```

File: frontend/workflows/sourcecontrol/package.json (hunk 5 above)
Comment (danielhochman) [reply frontend/workflows/sourcecontrol/package.json:16]: hmm i don't get why that would happen, null coalesce is part of ES2020

Comment (dschaller) [reply frontend/workflows/sourcecontrol/package.json:16]: yeah, I didn't dig into the project enough to figure out why this is an issue but it's definitely not transpiling it

**Comment by dschaller** [comment]
Huh I didn't see that behavior. What command did you run? `yarn compile:dev`?

**Comment by dschaller** [comment]
@danielhochman I think this is actually a problem even now with babel since I believe the underlying issue is within lerna. Can we merge this and then fix that in a subsequent PR?

**Review by danielhochman — APPROVED** [review]

**Pull request merged**